prow/jenkins/jenkins.go

```
@@ -401,14 +425,121 @@ func getJobName(spec *prowapi.ProwJobSpec) string {
+
+	if err := json.Unmarshal(data, &jobInfo); err != nil {
+		return nil, fmt.Errorf("Cannot unmarshal job info from API: %v", err)
+	}
+
+	c.logger.Tracef("JobInfo: %+v", jobInfo)
+
+	return &jobInfo, nil
+}
+
+// JobParameterized tells us if the Jenkins job for this ProwJob is parameterized
+func (c *Client) JobParameterized(jobInfo *JobInfo) bool {
+	for _, prop := range jobInfo.Property {
+		if prop.ParameterDefinitions != nil && len(prop.ParameterDefinitions) > 0 {
+			return true
+		}
+	}
+
+	return false
+}
+
+// EnsureBuildableJob attempts to detect a job that hasn't yet ran and populated
+// its parameters. If detected, it tries to run a build until the job parameters
+// are processed, then it aborts the build.
+func (c *Client) EnsureBuildableJob(spec *prowapi.ProwJobSpec) error {
+	jobInfo, err := c.GetJobInfo(spec)
+
+	if err != nil {
+		return err
+	}
+
+	isParameterized := c.JobParameterized(jobInfo)
+
+	c.logger.Tracef("JobHasParameters: %v", isParameterized)
+
+	if isParameterized || len(jobInfo.Builds) > 0 || err != nil {
+		return err
+	}
+
+	path := getBuildPath(spec)
+	c.logger.Debugf("getBuildPath: %s", path)
+
+	resp, err := c.request(http.MethodPost, path, nil, true)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 201 {
+		return fmt.Errorf("EnsureBuildableJob build job response not 201: %s", resp.Status)
+	}
+
+	backoff := wait.Backoff{
+		Duration: time.Duration(5) * time.Second,
+		Factor:   1,
+		Jitter:   1,
+		Steps:    10,
+	}
+
+	return wait.ExponentialBackoff(backoff, func() (bool, error) {
+		c.logger.Debugf("Waiting for job %v to become parameterized", spec.Job)
+
+		jobInfo, _ := c.GetJobInfo(spec)
+		isParameterized := false
+
+		if jobInfo != nil {
+			isParameterized = c.JobParameterized(jobInfo)
+
+			if isParameterized && jobInfo.LastBuild != nil {
+				c.logger.Debugf("Job %v is now parameterized, aborting the build", spec.Job)
+				c.Abort(getJobName(spec), &jobInfo.LastBuild)
```

does this ignore the error? should at least log it

---

I figured it's not worth failing the whole thing if for some reason the previous build wasn't aborted, certainly not against logging the error.